File: samplejava/src/main/java/pyxis/uzuki/live/richutils/samplejava/RefreshRecyclerActivity.java
package pyxis.uzuki.live.richutils.samplejava;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;

import butterknife.BindView;
import kotlin.Unit;
import kotlin.jvm.functions.Function0;
import pyxis.uzuki.live.richutilskt.utils.Utils;
import pyxis.uzuki.live.richutilskt.widget.RefreshRecyclerView;

/**
 * Created by winds on 2017-07-03.
 */

public class RefreshRecyclerActivity extends BaseActivity {
    @BindView(R.id.recyclerView) RefreshRecyclerView recyclerView;
    private ArrayList<String> itemSet = new ArrayList<>();
    private int finalLoaded = 0;
    private ListAdapter adapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_refresh_recycler);

        recyclerView.setOnRefreshListener(() -> {
            for (int i = finalLoaded; i < finalLoaded + 9; i++) {
                itemSet.add(String.valueOf(i));
            }

            finalLoaded += 9;
            adapter.notifyDataSetChanged();
            return Unit.INSTANCE;
        });

        adapter = new ListAdapter();
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adapter);

        for (int i = 0; i < 20; i++) {
            itemSet.add(String.valueOf(i));
        }

        adapter.notifyDataSetChanged();
    }

    private class ListAdapter extends RecyclerView.Adapter<ListHolder> {

        @Override
        public ListHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
            return new ListHolder(Utils.inflate(viewGroup.getContext(), R.layout.activity_refresh_recycler_item, viewGroup));
        }

        @SuppressLint("SetTextI18n")
        @Override
        public void onBindViewHolder(ListHolder listHolder, int i) {
            listHolder.txtItem.setText("This is great code! -> " + i);
        }

        @Override
        public int getItemCount() {
            return itemSet.size();
        }
    }

    private class ListHolder extends RecyclerView.ViewHolder {
        private TextView txtItem;

        ListHolder(View itemView) {
            super(itemView);
            txtItem = (TextView) itemView.findViewById(R.id.txtItem);
        }
    }
}

File: sample/src/main/java/pyxis/uzuki/live/richutilssample/RefreshRecyclerActivity.kt
package pyxis.uzuki.live.richutilssample

import android.annotation.SuppressLint
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.activity_refresh_recycler.*
import kotlinx.android.synthetic.main.activity_refresh_recycler_item.view.*
import pyxis.uzuki.live.richutilskt.inflate
import pyxis.uzuki.live.richutilskt.toast

class RefreshRecyclerActivity : AppCompatActivity() {
    private var itemSet: ArrayList<String> = ArrayList()
    private var finalLoaded = 0
    private var adapter: ListAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_refresh_recycler)

        recyclerView.setOnRefreshListener {
            for (i in finalLoaded..finalLoaded + 9) {
                itemSet.add(i.toString())
            }
            finalLoaded += 9
            adapter?.notifyDataSetChanged()
        }

        adapter = ListAdapter { position: Int, text: String ->
            toast("clicked ${position}nd item with {$text}")
        }

        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = adapter

        for (i in 0..20) {
            itemSet.add(i.toString())
        }

        (adapter as ListAdapter).notifyDataSetChanged()
    }

    private inner class ListAdapter(var itemClick: (Int, String) -> Unit) : RecyclerView.Adapter<ListHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListHolder =
                ListHolder(parent.context.inflate(R.layout.activity_refresh_recycler_item, parent), itemClick = itemClick)

        override fun onBindViewHolder(holder: ListHolder, position: Int) = holder.bind()
        override fun getItemCount(): Int = itemSet.size
    }

    private inner class ListHolder(itemView: View, var itemClick: (Int, String) -> Unit) : RecyclerView.ViewHolder(itemView) {

        @SuppressLint("SetTextI18n")
        fun bind() {
            itemView.txtItem.text = "This is great code! -> $layoutPosition"
            itemView.setOnClickListener { itemClick(layoutPosition, itemView.txtItem.text.toString()) }
        }
    }
}